### PR TITLE
Separate server ID from location for controller

### PR DESCRIPTION
**Pull request by NotTheEvilOne**
This PR adds support to execute the CSI controller on a node not hosted on the Hetzner Cloud. Additionally it reduces required API calls in this case as the location is already known at execution.

It adds a new environment variable `HCLOUD_SERVER_OVERRIDE_LOCATION_NAME` used for the controller service to work properly even without a server ID.

This is useful in a Gardener deployment where the "csi-driver-controller" is not necessarily deployed on a node in the Hetzner Cloud offering but still manages volumes for nodes deployed on the cloud.

### CODE REVIEW

**Comment by NotTheEvilOne** [comment]
Is there any news regarding this repository @LKaemmerling or @fhofherr ?

**Comment by apricote** [comment]
It looks like this location is only used once: https://github.com/hetznercloud/csi-driver/blob/c83b6cc77a7f890a40095cdfd762e001badd602c/driver/controller.go#L57-L63

Considering this, I wonder If we can just completely remove the code, as all volumes should have topology requirements attached by now (GA in Kubernetes since 1.17).

**Comment by apricote** [comment]
> It looks like this location is only used once:
> 
> https://github.com/hetznercloud/csi-driver/blob/c83b6cc77a7f890a40095cdfd762e001badd602c/driver/controller.go#L57-L63
> 
> Considering this, I wonder If we can just completely remove the code, as all volumes should have topology requirements attached by now (GA in Kubernetes since 1.17).

While this is (IMO) technically allowed according to the CSI spec, the `kubernetes-csi/csi-test` sanity tests fail if we require the accessibilityRequirements to always be set.

I researched how other csi-drivers are doing this, and some also accept config options to override the default location (`digitalocean` and `gce-persistent-disk`).

File: cmd/controller/main.go [hunk 1]
```diff
@@ -30,14 +30,21 @@ func main() {
...
-	if err != nil {
-		level.Error(logger).Log(
-			"msg", "failed to fetch server",
-			"err", err,
-		)
-		os.Exit(1)
+	if s := os.Getenv("HCLOUD_SERVER_OVERRIDE_LOCATION_NAME"); s != "" {
+		location = s
+	} else {
+		server, err := app.GetServer(logger, hcloudClient, metadataClient)
```
Comment (apricote) [inline cmd/controller/main.go:38]: `metadataClient` is only used in this branch. Can you move the `metadataClient := metadata.NewClient()` line here?

File: cmd/controller/main.go [hunk 2]
```diff
@@ -30,14 +30,21 @@ func main() {
...
+	var location string
 
-	server, err := app.GetServer(logger, hcloudClient, metadataClient)
-	if err != nil {
-		level.Error(logger).Log(
-			"msg", "failed to fetch server",
-			"err", err,
-		)
-		os.Exit(1)
+	if s := os.Getenv("HCLOUD_SERVER_OVERRIDE_LOCATION_NAME"); s != "" {
```
Comment (apricote) [inline cmd/controller/main.go:35]: What do you think about calling this var `HCLOUD_VOLUME_DEFAULT_LOCATION` instead?

While this technically provides an override to the server location check, the returned value is used as the default location for volumes which have no location specified.

**Review by apricote — CHANGES_REQUESTED** [review]

**Comment by apricote** [comment]
Hey @NotTheEvilOne, 

I saw that you created 23technologies/hcloud-csi-driver#1 with the requested changes. Should we wait for that PR or just create a new one here with the contents?

**Comment by NotTheEvilOne** [comment]
Good morning,

I guess we should just go ahead and take over the changes from my [user repository](https://github.com/hetznercloud/csi-driver/compare/main...NotTheEvilOne:hcloud-csi-driver:patch-1). Please let me know if you want me to create a new PR for it.

**Comment by apricote** [comment]
> Good morning,
> 
> I guess we should just go ahead and take over the changes from my [user repository](https://github.com/hetznercloud/csi-driver/compare/main...NotTheEvilOne:hcloud-csi-driver:patch-1). Please let me know if you want me to create a new PR for it.

That would be great :)

**Comment by NotTheEvilOne** [comment]
Superseded by #385.

**Comment by apricote** [comment]
Replaced by #385.

**Pull request closed without merging**